pathway-analysis/02-gene_set_enrichment_analysis.Rmd

```diff
@@ -111,77 +107,143 @@ dge_results_df <- readr::read_tsv(dge_results_file)
+  dplyr::pull(gene_symbol)
 ```
 
-### Preranked list
+Now we'll look at the values for the 
 
-The `GSEA()` function takes a preranked (sorted) named vector of statistics, where the names in the vector are gene identifiers.
+```{r show_gene_dups}
+dge_results_df %>%
+  dplyr::filter(gene_symbol %in% duplicated_gene_symbols) %>%
+  dplyr::arrange(gene_symbol)
+```
+We can see that the associated values vary for each row.
+
+Let's keep the gene symbols associated with the higher absolute value of the log2 fold change.
+GSEA relies on genes' rankings on the basis of a gene-level statistic and the enrichment score that is calculated reflects the degree to which genes in a gene set are overrepresented in the top or bottom of the rankings ([Subramanian, Tamayo *et al.* 2005](https://doi.org/10.1073/pnas.0506580102); [Yu](http://yulab-smu.top/clusterProfiler-book/chapter2.html#gene-set-enrichment-analysis)).
+
+Retaining the instance of the gene symbols with the higher absolute value of a gene-level statistic means that we will retain the value that is likely to be more highly- or lowly-ranked or, put another way, the values less likely to be towards the middle of the ranked gene list.
+We should keep this decision in mind when interpreting our results.
+For example, if all the duplicate identifiers happened to be in a particular gene set, we may get an overly optimistic view of how perturbed that gene set is because we preferentially selected instances of the identifier that have a higher absolute value of the statistic used for ranking.
+
+In the next chunk, we are going to filter out the duplicated row using the `dplyr::distinct()` function
+This will keep the first row with the duplicated value thus keeping the row with the highest absolute value of the log2 fold change.
+
+```{r filter_dge}
+filtered_dge_df <- dge_results_df %>%
+  # Sort so that the highest absolute values of the log2 fold change are at the
+  # top
+  dplyr::arrange(dplyr::desc(abs(log2FoldChange))) %>%
+  # Filter out the duplicated rows using `dplyr::distinct()`
+  dplyr::distinct(gene_symbol, .keep_all = TRUE)
+```
+
+Let's see what happened to our duplicate identifiers.
+
+```{r show_filtered_dge, live = TRUE}
+# Subset to & arrange by gene symbols that were duplicated in the original 
+# data frame of results
+filtered_dge_df %>%
+  dplyr::filter(gene_symbol %in% duplicated_gene_symbols) %>%
+  dplyr::arrange(gene_symbol)
+```
+
+Now we're ready to prep our pre-ranked list for GSEA.
+
+### Pre-ranked list
+
+The `GSEA()` function takes a pre-ranked (sorted) named vector of statistics, where the names in the vector are gene identifiers.
 This is step 1 -- gene-level statistics.
 
 ```{r lfc_vector}
-lfc_vector <- dge_results_df$log2FoldChange
-names(lfc_vector) <- dge_results_df$entrez_id
+lfc_vector <- filtered_dge_df$log2FoldChange
+names(lfc_vector) <- filtered_dge_df$gene_symbol
 lfc_vector <- sort(lfc_vector, decreasing = TRUE)
```

Comment:
Because I hate `names() <-` (and generally assigning to a function), I'll suggest doing this with a newish feature of `dplyr::pull()`

```suggestion
lfc_vector <- filtered_dge_df%>%
  dplyr::pull(log2FoldChange, names = gene_symbol)
lfc_vector <- sort(lfc_vector, decreasing = TRUE)
```

Could also do the following, but I don't know whether I would care one way or the other.
```
lfc_vector <- filtered_dge_df%>%
  dplyr::arrange(dplyr::desc(log2FoldChange)) %>%
  dplyr::pull(log2FoldChange, names = gene_symbol)
```

Reply:
The argument is `name`, which makes less sense to me than `names`, but this is a note for future me and future you when we get a somewhat cryptic error when we use `names`.